File: Week1_story/story1.cpp
/**
 * @filename    - Story1.cpp
 * @description - The Efficient Traveler
 * @author      - Ehtesham Nawaz
 */
#include <iostream>
#include <vector>
using namespace std;

int EfficientTraveler(vector<int>& nums, int initialEnergy) {
    int energy = initialEnergy, farthest = -1;
    for (int i = 0; i < nums.size(); i++) {
        energy -= nums[i];
        if (energy < 0) break;
        farthest = i;
    }
    return farthest;
}

int main() {
    int n, energy;
    cout << "Enter number of cities: ";
    cin >> n;
    vector<int> nums(n);
    cout << "Enter energy costs: ";
    for (int i = 0; i < n; i++) cin >> nums[i];
    cout << "Enter initial energy: ";
    cin >> energy;
    cout << "Farthest city reachable: " << EfficientTraveler(nums, energy) << endl;
}

File: Week_1/Story1.java
import java.util.*;
public class Story1{

    public static int farthestCity(ArrayList<Long> energyCost, long initialEnergy) {

        long currentEnergy = initialEnergy;
        int totalCities = energyCost.size();

        for(int cityIndex = 0; cityIndex < totalCities; cityIndex++) {

            currentEnergy -= energyCost.get(cityIndex);

            if(currentEnergy < 0) {
                return cityIndex - 1;
            }
        }

        return totalCities;
    }

    public static void main(String[] args) {

        Scanner sc = new Scanner(System.in);

        System.out.print("Enter number of cities: ");
        int numberOfCities = sc.nextInt();
        ArrayList<Long> energyCost = new ArrayList<>();

        System.out.println("Enter the energy cost for each city:");
        for(int i = 0; i < numberOfCities; i++) {
            energyCost.add(sc.nextLong());
        }

        System.out.print("Enter initial energy: ");
        long initialEnergy = sc.nextLong();

        int answer = farthestCity(energyCost, initialEnergy);

        System.out.println("Farthest reachable city index = " + answer);
    }
}
